python/python/res/simulator/robust_simulator.py

```diff
@@ -0,0 +1,121 @@
+from ecl.util import BoolVector
+
+from res.enkf import ErtRunContext, EnKFMain, EnkfConfigNode, EnkfNode, NodeId
+from res.server import SimulationContext
+from .robust_simulator_context import RobustContext
+
+class RobustSimulator(object):
+
+    def __init__(self, res_config, controls, results):
+        """Will create simulator which can be used to run multiple simulations.
+
+        The @res_config argument should be a ResConfig object, representing the
+        fully configured state of libres.
+
+
+        The @controls argument configures which parameters the simulator should
+        get when actually simulating. The @controls argument should be a
+        dictianry like this:
+
+            controls = {"cmode": ["Well","Group"], "order" : ["W1", "W2", "W3"]}
+
+        In this example the simulator will expect two arrays 'cmode' and
+        'order', consisting of two and three elements respectively. When
+        actually simualating these values will be written to json files looking
+        like:
+
+             cmode.json = {"Well" : 1.0, "Group" : 2.0}
+             order.json = {"W1" : 1, "W2" : 1.0, "W3": 1.0}
+
+        When later invoking the start() method the simulator expects to get
+        values for all parameters configured with the @controls argument,
+        otherwise an exception will be raised. Internally in libres code the
+        controls will be implemented as 'ext_param' instances.
+
+
+        The @results argument is a list of keys of results which the simulator
+        expects to be generated by the forward model. If argument @results
+        looks like:
+
+             results = ["CMODE", "order"]
+
+        The simulator will look for the files 'CMODE_0' and 'order_0' in the
+        simulation folder. If those files are not produced by the simulator an
+        exception will be raised.
+        """
+
+        self.res_config = res_config
+        self.ert = EnKFMain( self.res_config )
+        self.control_keys = []
+        self.result_keys = []
+
+        ens_config = self.res_config.ensemble_config
+        for key in controls.keys():
+            ens_config.addNode( EnkfConfigNode.create_ext_param( key , controls[key] ))
+            self.control_keys.append(key)
```

Comment:
Just a minor comment, no need to change the code here. I am myself unsure about the following, so I take the chance to discuss it.

If you do ``` for k in d.keys: print(d[k])```, then I guess `k` has to be looked up in the dictionary at every iteration, which is potentially slow

I contrast, I would expect ```for k, v in d.iteritems(): print(v)``` to be faster since it doesn't have to lookup the value at every iteration (and I think it looks better).

Has anyone strong opinions about this? Is there any well-defined "best practice" about this?

Reply:
> Has anyone strong opinions about this? Is there any well-defined "best practice" about this?

I certainly don't know - actually I feel that this is one corner of Python where the "One natural & good way to do it" does not really apply.

Reply:
I agree with both of you, I too find the later to look nicer, but it does appear to deviate from the _one way_ mentality.